### PR TITLE
[MRG] New context manager API for joblib.Parallel

### PR DESCRIPTION
**Pull request by ogrisel**
This should make it easier to use parallel calls interleaved in loop as done in scikit-learn/scikit-learn#4873 for instance (ping @arthurmensch).

### CODE REVIEW

**Comment by ogrisel** [comment]
@arthurmensch if you want to try this branch of joblib with scikit-learn you can git clone joblib, checkout this branch then run `pip install -e .` in this folder to install joblib in developer mode. Then create a `~/.pythonstartup.py` file with the following monkeypatching code:

```python
try:
    import joblib
    from sklearn.externals import joblib as skl_joblib
    print('Monkeypatching scikit-learn embedded joblib')
    for k, v in vars(joblib).items():
        setattr(skl_joblib, k, v)
    del skl_joblib, joblib
except ImportError:
    pass
```

**Comment by ogrisel** [comment]
@GaelVaroquaux @lesteve you might want to have a quick look at this. Esp. the tests and the doc if you are in a hurry.

File: joblib/parallel.py [hunk 1]
```diff
@@ -441,12 +440,110 @@ def __init__(self, n_jobs=1, backend='multiprocessing', verbose=0,
...
+        # unless they choose to use the context manager API with a with block.
+        self._pool = None
         self._output = None
         self._jobs = list()
-        # A flag used to abort the dispatching of jobs in case an
-        # exception is found
-        self._aborting = False
+        self._managed_pool = False
+
+        # This lock is use to protect to coordinate the master thread of this
```
Comment (lesteve) [inline joblib/parallel.py]: "to protect to coordinate". Did you mean to only use one of the two verbs?

File: joblib/parallel.py [hunk 2]
```diff
@@ -441,12 +440,110 @@ def __init__(self, n_jobs=1, backend='multiprocessing', verbose=0,
...
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._terminate_pool()
+        self._managed_pool = False
+
+    def _effective_n_jobs(self):
+        n_jobs = self.n_jobs
+        if n_jobs == 0:
+            raise ValueError('n_jobs == 0 in Parallel has no meaning')
+        elif mp is None or n_jobs is None:
+            # multiprocessing is not avaiable or disabled, fallback
```
Comment (lesteve) [inline joblib/parallel.py]: typo: avai**l**able

File: joblib/test/test_parallel.py [hunk 3]
```diff
@@ -180,29 +208,52 @@ def test_error_capture():
...
+            # The managed pool should still be available and be in a working
+            # state despite the previsouly raised (and caught) exception
+            assert_true(parallel._pool is not None)
+            assert_equal([f(x, y=1) for x in range(10)],
+                         parallel(delayed(f)(x, y=1) for x in range(10)))
+
+            assert_raises(WorkerInterrupt, parallel,
+                          [delayed(interrupt_raiser)(x) for x in (1, 0)])
+
+            # The pool should still be avaible despite the exection
```
Comment (lesteve) [inline joblib/test/test_parallel.py]: typo: available
typo: exception

File: joblib/parallel.py [hunk 4]
```diff
@@ -441,12 +440,110 @@ def __init__(self, n_jobs=1, backend='multiprocessing', verbose=0,
...
+
+                # We are using multiprocessing, we also want to capture
+                # KeyboardInterrupts
+                self.exceptions.extend([KeyboardInterrupt, WorkerInterrupt])
+        else:
+            raise ValueError("Unsupported backend: %s" % self.backend)
+        return n_jobs
+
+    def _terminate_pool(self):
+        pool = getattr(self, '_pool', None)
```
Comment (lesteve) [inline joblib/parallel.py]: `self._pool` is defined in `__init__` so you don't need the getattr I think.

Comment (ogrisel) [inline joblib/parallel.py]: Indeed.

**Comment by ogrisel** [comment]
@lesteve all comments are addressed and appveyor has run the tests under windows as well. Once merged I think we should release beta3 and sync sklearn to get the opportunity to test it.

**Comment by lesteve** [comment]
Merging, thanks!

**Comment by lesteve** [comment]
I'll do a release and open a PR in scikit-learn.

**Comment by GaelVaroquaux** [comment]
> Merging, thanks!

Awesome!

**Comment by ogrisel** [comment]
:beers:

**Pull request merged**